language/tools/move-cli/src/sandbox/commands/publish.rs

```diff
@@ -3,118 +3,90 @@
-        .set_flags(Flags::empty().set_sources_shadow_deps(republish))
-        .set_named_address_values(named_address_mapping.clone())
-        .build_and_report()?;
-
-    let num_modules = compiled_units
-        .iter()
-        .filter(|u| matches!(u, AnnotatedCompiledUnit::Module(_)))
-        .count();
-    if verbose {
-        println!("Found and compiled {} modules", num_modules)
+        println!(
+            "Found {} modules",
+            package.modules().collect::<Vec<_>>().len()
+        );
     }
 
-    let mut modules = vec![];
-    for c in compiled_units {
-        match c {
-            AnnotatedCompiledUnit::Script(_) => {
-                if verbose {
-                    println!(
-                        "Warning: Found script in specified files for publishing. But scripts \
-                         cannot be published. Script found in: {}",
-                        c.loc().file_hash()
-                    )
+    if no_republish {
+        let republished = package
+            .modules()
+            .filter_map(|unit| {
+                let id = module(&unit.unit).ok()?.self_id();
+                if state.has_module(&id) {
+                    Some(format!("{}", id))
+                } else {
+                    None
                 }
-            }
-            AnnotatedCompiledUnit::Module(annot_module) => modules.push((
-                (
-                    annot_module.module_ident(),
-                    annot_module.address_name.map(|n| n.value),
-                ),
-                annot_module.named_module.module,
-                annot_module.named_module.source_map,
-            )),
+            })
+            .collect::<Vec<_>>();
+
+        if !republished.is_empty() {
+            bail!("Failed to republish modules since the --no-republish flag is set. Tried to republish the following modules: {}",
+                republished.join(", "))
         }
     }
 
-    // use the the publish_module API frm the VM if we do not allow breaking changes
+    // use the the publish_module API from the VM if we do not allow breaking changes
     if !ignore_breaking_changes {
-        let id_to_ident: BTreeMap<_, _> = modules
-            .iter()
-            .map(|((_, addr_name_opt), module, _)| {
-                let id = module.self_id();
-                (id, *addr_name_opt)
-            })
-            .collect();
-
         let vm = MoveVM::new(natives).unwrap();
         let mut gas_status = get_gas_status(None)?;
         let mut session = vm.new_session(state);
 
         let mut has_error = false;
         match override_ordering {
             None => {
-                for (_, module, src_map) in &modules {
-                    let mut module_bytes = vec![];
-                    module.serialize(&mut module_bytes)?;
-
-                    let id = module.self_id();
+                for unit in package.modules() {
```

Comment:
Somewhat of a pre-existing issue, but I think you want to use the VM's multi-module publish API here rather than publishing modules one-by-one. Otherwise, packages that use `friend` will fail to publish with a cryptic `LINKER_ERROR`.

Reply:
Good point -- let me put this up in separate PR, as there are some intricacies that need to be worked out here as well, e.g., if the package has multiple friend bundles with different addresses mixed in with non-friend modules in different addresses that may depend on these bundles, the publishing order becomes harder.